spec/fixtures/generate_external_certs.sh

```diff
@@ -0,0 +1,337 @@
+  
+  FILE=/tmp/ca_gen/site_$1.req.conf
+  site_request_conf $FILE $4
+  if [ "$7" != "0" ]; then append_altnames    $7 $FILE; fi
+  if [ "$8" != "0" ]; then append_ip_altnames $8 $FILE; fi
+  
+  openssl req -new -sha256 -key $DIR_KEYS/site_$1.key -subj "$6" -out $CSR_DIR/site_$1.csr -config $FILE
+  # It is possible to specify altnames the following way instead of constructing a conf file.
+  # There are additional ways beyond this using new features of latest openssl as well.
+  #openssl req -new -sha256 \
+  #  -key keys/site_$1.key \
+  #  -subj "$6" \
+  #  -reqexts SAN \
+  #  -config <(cat /etc/ssl/openssl.cnf \
+  #   <(printf "\n[SAN]\nsubjectAltName=DNS:example.com,DNS:www.example.com")) \
+  #  -out $CSR_DIR/site_$1.csr
+  
+  # Cleanest way to see the actual contents of the CSR to verify it was created correctly.
+  #openssl req -in $CSR_DIR/site_$1.csr -text -noout
+  
+  # Unclean very raw way to see the contents of the CSR
+  # openssl asn1parse < $CSR_DIR/site_$1.csr
+  
+  gen_start_end $5
+  
+  # Irritatingly openssl ca command does not, by default, use the altnames specified in the CSR.
+  # It ignores them. They have to be re-specified. It may be pointless to specify them in the CSR as a result.
+  # The altnames can be copied from the request potentially, by using the "copy_extensions = copy" option of openssl conf
+  cp $DIR_CONF/ca_$2.conf $DIR_CONF/site_$1.conf
+  if [ "$7" != "0" ]; then append_altnames    $7 $DIR_CONF/site_$1.conf; fi
+  if [ "$8" != "0" ]; then append_ip_altnames $8 $DIR_CONF/site_$1.conf; fi
+  
+  openssl ca -batch \
+    -config $DIR_CONF/site_$1.conf \
+    -extensions req_ext \
+    -out $DIR_CERTS/site_$1.crt \
+    -md $3 \
+    -startdate $START_DATE -enddate $END_DATE \
+    -infiles $CSR_DIR/site_$1.csr
+}
+
+function create_site_crt_selfsigned() { # $1 domain, $2 message_digest, $3 rsa_bits, $4 good/expired, $5 subject, $6 altnames, $7 ipaltnames
+  echo "Generating site key \"$1\""
+  openssl genrsa -out $DIR_KEYS/site_$1.key $3 2>/dev/null
+  
+  init_root_ca dummy "/CN=US/O=US/OU=US Unit"
+
+  FILE=/tmp/ca_gen/site_$1.req.conf
+  site_request_conf $FILE $3
+  if [ "$6" != "0" ]; then append_altnames    $6 $FILE; fi
+  if [ "$7" != "0" ]; then append_ip_altnames $7 $FILE; fi
+  
+  openssl rsa -in $DIR_KEYS/site_$1.key -out $DIR_KEYS/site_rsa_$1.key
+  openssl req -new -sha256 -key $DIR_KEYS/site_$1.key -subj "$5" -out $CSR_DIR/site_$1.csr -config $FILE
+
+  cp $DIR_CONF/ca_dummy.conf $DIR_CONF/site_$1.conf
+  
+  # Generate an extensions file; not extending default config since x509 command doesn't have a way to specify a
+  # full config file that I see.
+  #echo -e "[req_ext]\nsubjectAltName = @alt_names\n" >> conf/site_$1.conf
+  #echo -e "[alt_names]" >> conf/site_$1.conf
+  if [ "$6" != "0" ]; then append_altnames    $6 $DIR_CONF/site_$1.conf; fi
+  if [ "$7" != "0" ]; then append_ip_altnames $7 $DIR_CONF/site_$1.conf; fi
+  
+  gen_start_end $4
+  
+  #openssl x509 -req \
```

Comment:
This method has issues. It is not explained clearly why it is not used here and this commented out block should just be removed as a result. It was left initially because it seemed to be the desirable and suggested way to do it; but it has problems...

Reply:
Removed the commented out portion and fixed the comment above it.